README.md
# Real email validation for Laravel

[![Build Status](https://img.shields.io/travis/InteractionDesignFoundation/laravel-real-email-validation/master.svg?style=flat-square)](https://travis-ci.org/InteractionDesignFoundation/laravel-real-email-validation)
[![Code coverage](https://scrutinizer-ci.com/g/InteractionDesignFoundation/laravel-real-email-validation/badges/coverage.png)](https://scrutinizer-ci.com/g/InteractionDesignFoundation/laravel-real-email-validation)
[![Quality Score](https://img.shields.io/scrutinizer/g/InteractionDesignFoundation/laravel-real-email-validation.svg?style=flat-square)](https://scrutinizer-ci.com/g/InteractionDesignFoundation/laravel-real-email-validation)
[![StyleCI](https://github.styleci.io/repos/200292916/shield?branch=master)](https://github.styleci.io/repos/200292916)
[![Latest Version on Packagist](https://img.shields.io/packagist/v/idf/laravel-real-email-validation.svg?style=flat-square)](https://packagist.org/packages/idf/laravel-real-email-validation)
[![Total Downloads](https://img.shields.io/packagist/dt/idf/laravel-real-email-validation.svg?style=flat-square)](https://packagist.org/packages/idf/laravel-real-email-validation)

Laravel has a good `email` validation rule, but it can miss some invalid email addresses.
This packages aims to cover more cases.

## Installation

You can install the package via composer:

```bash
composer require idf/laravel-real-email-validation
```

The package will automatically register itself.

### Translations

If you wish to edit the package translations, you can run the following command to publish them into your `resources/lang` folder

```bash
php artisan vendor:publish --provider="IDF\RealEmailValidation\ServiceProvider"
```

## Usage

```php
$this->validate($request, [
    'email' => ['required', 'email', new RealEmail()],
]);
```

By default it uses the following checks: `html5`, `rfc`, `host`, but you can define your set:
```php
// checks without network requests
new RealEmail(['html5', 'rfc'])
```

1. `html5`: Uses regex pattern for rules defined by [WHATWG](https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address). Browsers use it for `input[type="email"]`.
1. `rfc`: Strict RFC validation. Check against RFC 5321, 5322, 6530, 6531, 6532, treats warnings as errors.
1. `host`: Checks DNS Records for the host extracted from email address. Uses network.
1. `mx`: Check DNS Records for MX type only. ⚠️ This option is not reliable because it depends on the network conditions and some valid servers refuse to respond to those requests.

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

### Security

If you discover any security related issues, please email hello@team.interaction-design.org instead of using the issue tracker.

## Credits

- [All Contributors](../../contributors)
- Symfony validation (for inspiration)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
Changelog
All notable changes to `laravel-real-email-validation` will be documented in this file.


tests/TestCase.php
<?php

namespace IDF\RealEmailValidation\Tests;

use IDF\RealEmailValidation\ServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

abstract class TestCase extends Orchestra
{
    protected function getPackageProviders($app)
    {
        return [
            ServiceProvider::class,
        ];
    }
}


tests/Rules/RealEmailTest.php
<?php

namespace IDF\RealEmailValidation\Tests\Rules;

use Illuminate\Contracts\Validation\Rule;
use IDF\RealEmailValidation\Tests\TestCase;
use IDF\RealEmailValidation\Rules\RealEmail;

/**
 * Inspired by
 * https://github.com/symfony/validator/blob/4.4/Tests/Constraints/EmailValidatorTest.php.
 */
class RealEmailTest extends TestCase
{
    /**
     * @test
     * @dataProvider validEmailProvider
     * @group network
     */
    public function it_passes_on_valid_email($validEmail)
    {
        $rule = new RealEmail(['html5', 'rfc', 'host'], false);

        $this->assertTrue($rule->passes('random_input_name', $validEmail), $this->getErrorMessages($rule));
    }

    /**
     * @test
     * @dataProvider invalidEmailsForHostCheckProvider
     * @group network
     */
    public function it_fails_on_invalid_host($invalidEmail)
    {
        $rule = new RealEmail(['host']);

        $this->assertFalse($rule->passes('random_input_name', $invalidEmail));
    }

    /**
     * @test
     * @dataProvider invalidEmailForRfcCheckProvider
     */
    public function it_fails_by_rfc_check($invalidEmail)
    {
        $rule = new RealEmail(['rfc']);

        $this->assertFalse($rule->passes('random_input_name', $invalidEmail), $invalidEmail);
    }

    /**
     * @test
     * @dataProvider invalidEmailsForHtml5CheckProvider
     */
    public function it_fails_by_html5_check($invalidEmail)
    {
        $rule = new RealEmail(['html5']);

        $this->assertFalse($rule->passes('random_input_name', $invalidEmail), $invalidEmail);
    }

    public function validEmailProvider()
    {
        return [
            ['example@gmail.com'],
            ['example@yahoo.com'],
            ['example@example.co.uk'],
        ];
    }

    public function invalidEmailsForHostCheckProvider()
    {
        return [
            ['example@gmail.con'],
            ['example@gmail.ocm'],
        ];
    }

    public function invalidEmailForRfcCheckProvider()
    {
        return [
            ['test@example.com test'],
            ['user  name@example.com'],
            ['user   name@example.com'],
            ['example.@example.co.uk'],
            ['example@example@example.co.uk'],
            ['(test_exampel@example.fr)'],
            ['example(example)example@example.co.uk'],
            ['.example@localhost'],
            ['ex\ample@localhost'],
            ['example@local\host'],
            ['example@localhost.'],
            ['user name@example.com'],
            ['username@ example . com'],
            ['example@(fake).com'],
            ['example@(fake.com'],
            ['username@example,com'],
            ['usern,ame@example.com'],
            ['user[na]me@example.com'],
            ['"""@iana.org'],
            ['"\"@iana.org'],
            ['"test"test@iana.org'],
            ['"test""test"@iana.org'],
            ['"test"."test"@iana.org'],
            ['"test".test@iana.org'],
            ['"test"'.\chr(0).'@iana.org'],
            ['"test\"@iana.org'],
            [\chr(226).'@iana.org'],
            ['test@'.\chr(226).'.org'],
            ['\r\ntest@iana.org'],
            ['\r\n test@iana.org'],
            ['\r\n \r\ntest@iana.org'],
            ['\r\n \r\ntest@iana.org'],
            ['\r\n \r\n test@iana.org'],
            ['test@iana.org \r\n'],
            ['test@iana.org \r\n '],
            ['test@iana.org \r\n \r\n'],
            ['test@iana.org \r\n\r\n'],
            ['test@iana.org  \r\n\r\n '],
            ['test@iana/icann.org'],
            ['test@foo;bar.com'],
            ['test;123@foobar.com'],
            ['test@example..com'],
            ['email.email@email."'],
            ['test@email>'],
            ['test@email<'],
            ['test@email{'],
            [str_repeat('x', 254).'@example.com'], //email with warnings
        ];
    }

    public function invalidEmailsForHtml5CheckProvider()
    {
        return [
            ['example'],
            ['example@'],
            ['example@localhost'],
            ['example@example.co..uk'],
            ['foo@example.com bar'],
            ['example@example.'],
            ['example@.fr'],
            ['@example.com'],
            ['example@example.com;example@example.com'],
            ['example@.'],
            [' example@example.com'],
            ['example@ '],
            [' example@example.com '],
            [' example @example .com '],
            ['example@-example.com'],
            [sprintf('example@%s.com', str_repeat('a', 64))],
        ];
    }

    private function getErrorMessages(Rule $rule): string
    {
        return json_encode($rule->message());
    }
}


src/Rules/RealEmail.php
<?php

declare(strict_types=1);

namespace IDF\RealEmailValidation\Rules;

use Illuminate\Contracts\Validation\Rule;
use Egulias\EmailValidator\EmailValidator;
use Egulias\EmailValidator\Validation\NoRFCWarningsValidation;

/**
 * Inspired by
 * https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Validator/Constraints/EmailValidator.php.
 */
final class RealEmail implements Rule
{
    /** Check against RFC 5321, 5322, 6530, 6531, 6532 (no warnings allowed) */
    public const CONSTRAIN_STRICT_RFC = 'rfc';

    /** Check DNS Records for the extracted host */
    public const CONSTRAIN_HOST = 'host';

    /** Check DNS Records for MX type */
    public const CONSTRAIN_MX = 'mx';

    /**
     * Matches the pattern used for the HTML5 email input element.
     * @see https://html.spec.whatwg.org/multipage/forms.html#email-state-typeemail
     */
    public const CONSTRAIN_HTML5 = 'html5';

    /** @see https://html.spec.whatwg.org/multipage/input.html#valid-e-mail-address */
    private const PATTERN_HTML5 = '/^[a-zA-Z0-9.!#$%&\'*+\\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+$/';

    private $constraints = [];
    private $messages = [];
    private $bail;

    public function __construct(array $constraints = ['html5', 'rfc', 'host'], bool $bail = true)
    {
        $this->constraints = $constraints;
        $this->bail = $bail;
    }

    /** {@inheritdoc} */
    public function passes($attribute, $email): bool
    {
        $this->messages = [];

        if ($this->hasConstraint('html5') && ! $this->checkHtml5($email)) {
            $this->messages[] = __('realEmailValidation::messages.html5');
            if ($this->bail) {
                return false;
            }
        }

        if ($this->hasConstraint('rfc')) {
            $rfcValidator = new EmailValidator();
            $hasHtml5Error = ! $rfcValidator->isValid($email, new NoRFCWarningsValidation());
            if ($hasHtml5Error) {
                $this->messages[] = __('realEmailValidation::messages.rfc');
                if ($this->bail) {
                    return false;
                }
            }
        }

        $host = (string) substr($email, strrpos($email, '@') + 1);

        if ($this->hasConstraint('mx') && ! $this->checkMX($host)) {
            $this->messages[] = __('realEmailValidation::messages.mx');
            if ($this->bail) {
                return false;
            }
        }

        if ($this->hasConstraint('host') && ! $this->checkHost($host)) {
            $this->messages[] = __('realEmailValidation::messages.host');
            if ($this->bail) {
                return false;
            }
        }

        return count($this->messages) === 0;
    }

    /** {@inheritdoc} */
    public function message(): array
    {
        return $this->messages;
    }

    private function hasConstraint(string $constrain): bool
    {
        return in_array($constrain, $this->constraints, true);
    }

    /**
     * Check against HTML5 pattern.
     */
    private function checkHtml5(string $email): bool
    {
        return (bool) preg_match(self::PATTERN_HTML5, $email);
    }

    /**
     * Check DNS Records for MX type.
     * ⚠ This option is not reliable because it depends on the network conditions
     * and some valid servers refuse to respond to those requests.
     */
    private function checkMX(string $host): bool
    {
        return $host !== '' && checkdnsrr($host, 'MX');
    }

    /**
     * Check if one of MX, A or AAAA DNS RR exists.
     */
    private function checkHost(string $host): bool
    {
        return $host !== ''
            && ($this->checkMX($host) || (checkdnsrr($host, 'A') || checkdnsrr($host, 'AAAA')));
    }
}
